### ISSUE TITLE
Add support for nestable ANSI sequences

### ISSUE THREAD

**Issue by bobtiernay-okta** [remkop/picocli#550]
It would be very convenient for CLI authors to have the ability to nest ANSI escape sequences which currently [isn't possible](https://picocli.info/#_ansi_colors_and_styles#_usage_help_with_styles_and_colors)  with picocli.

> Markup styles cannot be nested, for example: @|bold this @|underline that|@|@ will not work. You can achieve the same by combining styles, for example: @|bold this|@ @|bold,underline that|@ will work fine.

However, libraries such as [doowb/ansi-colors](https://github.com/doowb/ansi-colors) support such a mechanism. This works by keeping a stack of modifiers / colors when parsing (see [index.js](https://github.com/doowb/ansi-colors/blob/master/index.js) for the implementation).

In picocli, this could be added to support the following:

```
@|red This is @|bold very|@ red|@
```
 which would be semantically equivalent to:
```
@|red This is |@@|red,bold very|@@|red  red|@
```

**Comment by remkop** [comment]
I won't have time to work on this for the foreseeable future but I will consider pull requests to accomplish this that don't add too much code and have good unit tests.

**Comment by bobtiernay-okta** [comment]
Before creating tests and submitting a PR, would something like this be considered "too much code"?

```java
    public static String flatten(String text) {
        class Frame {
            String fg;
            String bg;
            Set<String> modifiers = new HashSet<>();

            String format() {
                String styles = "";
                if (fg != null) styles += fg;
                if (bg != null) styles += (!styles.isEmpty() ? "," : "") + bg;
                if (!modifiers.isEmpty())
                    for (String modifier : modifiers) styles += (!styles.isEmpty() ? "," : "") + modifier;

                return styles;
            }

            void merge(Frame previous) {
                if (fg == null) fg = previous.fg;
                if (bg == null) bg = previous.bg;
                modifiers.addAll(previous.modifiers);
            }
        }

        String flattened = "";
        Stack<Frame> frames = new Stack<>();
        Matcher matcher = Pattern.compile("(@\\|(\\S+))|(\\|@)|(.(?!@\\||\\|@)*)").matcher(text);
        while (matcher.find()) {
            if (matcher.group(1) != null) {
                Frame frame = new Frame();
                for (String style : matcher.group(2).split(","))
                    if (style.matches("bold|faint|italic|underline|blink|reverse|reset")) frame.modifiers.add(style);
                    else if (style.matches("bg\\([^)]+\\)")) frame.bg = style;
                    else frame.fg = style;

                if (!frames.isEmpty()) {
                    frame.merge(frames.peek());
                }

                flattened += (frames.isEmpty() ? "" : "|@") + "@|" + frame.format();
                frames.push(frame);
            } else if (matcher.group(3) != null) {
                frames.pop();
                flattened += frames.isEmpty() ? "|@" : "|@@|" + frames.peek().format();
            } else {
                flattened += matcher.group(4);
            }
        }

        return flattened;
    }
```

**Comment by remkop** [comment]
That’s about as much code as the current parsing logic, that’s ok. 

Before we proceed though, some questions: 
- is the goal to support nesting to arbitrary depth? Does your proposal accomplish that?
- what do we do if opening and closing symbols are unmatched?
- is it possible to show the string `@|` or `|@` in the text by escaping?

(Away from PC, didn’t check the answers for the current code for 2 and 3...)

**Comment by bobtiernay-okta** [comment]
> is the goal to support nesting to arbitrary depth? Does your proposal accomplish that?

Yes to both.

> what do we do if opening and closing symbols are unmatched

Similar to current behavior, it simply doesn't translate the symbols properly:

```java
flatten("Did you know that @|red this is @|bold,underline very|@ red, but @|italic this|@ isn't?");
```
```
Did you know that @|red this is |@@|red,underline,bold very|@@|red  red, but |@@|red,italic this|@@|red  isn't?
```

> is it possible to show the string @| or |@ in the text by escaping?

Hmm, I don't think so but we could support if needed. What' the current behavior?

As another point, I think we can remove the need for the regexes, but it might require a bit more code.

Lastly, I think the key to point of all of this is location transparency. In the current model, it is very challenging to encapsulate standard formatting, but allow for customization at the call site. The reason is the current formatting scheme is not composable. By allowing nesting, we make building more consistent color schemes possible.

**Comment by remkop** [comment]
I can see the benefits of allowing nesting. Composition is powerful. 

About the implementation:  note that `Help.Ansi.Style` has a method `parse(String)` that takes a comma-separated list of style strings (e.g., `"red,underline,bold,bg(yellow)"`) and returns an array of `IStyle` objects. Would that help save some work?

**Comment by bobtiernay-okta** [comment]
I actually tried that at first. However, since there is no `toString` method for that object, I cannot go the other way. Also note that it returns an `IStyle`.

**Comment by remkop** [comment]
Would an additional method `String toString(IStyle[])` help?

**Comment by bobtiernay-okta** [comment]
It would help slightly.

I think for maximal code collapse we would need to in-line this into the main parsing code. This would have the benefit of improving efficiency (avoiding regexes and multiple passes). 

If that approach is preferable I can take a stab at it when I get a chance. In the meantime, I’ve added the above implementation as a preprocessing step in my project to start giving it some real world variation. Seems to be working nicely after a few tweaks.

**Comment by bobtiernay-okta** [comment]
I don't know if this will add a lot of additional complexities to what is already there. The nice thing about it is that it will pipeline the process and leverage the already existing parse context which I need to duplicate for the straight flattening approach. 

That said, I think I would need to try it out. If it gets too hairy, then I agree its best to keep these facilities separate.

**Comment by remkop** [comment]
Makes sense. The `substring` and `concat` methods in `Text` were especially tricky to get right. It wouldn’t surprise me if the current implementation still has bugs.
